### PR TITLE
Added "Tensor tol" overload of torch.linalg.matrix_rank

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Currently `torch.linalg.matrix_rank` accepts only Python's float for `tol=` argument. The current behavior is not NumPy compatible and this PR adds the possibility to pass Tensor for matrix-wise tolerances.

Ref. #42666

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -196,53 +196,99 @@ Tensor pinverse(const Tensor& self, double rcond) {
...
+  checkSameDevice("torch.linalg.matrix_rank", rtol, input, "rtol");
   ScalarType output_type = ScalarType::Long;
-  checkLinalgCompatibleDtype("linalg_matrix_rank", result.scalar_type(), output_type);
+  checkLinalgCompatibleDtype("torch.linalg.matrix_rank", result.scalar_type(), output_type);
 
   // Matrices or batch of matrices are allowed
-  TORCH_CHECK(self.dim() >= 2, "linalg_matrix_rank: Expected as input a matrix or a batch of matrices, but got a tensor of size: ", self.sizes());
+  TORCH_CHECK(input.dim() >= 2, "torch.linalg.matrix_rank: Expected as input a matrix or a batch of matrices, but got a tensor of size: ", input.sizes());
+
+  TORCH_CHECK(!at::isComplexType(atol.scalar_type()),
```
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:213]: Why not restrict atol and rtol to be the same dtype as input?

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -196,53 +196,99 @@ Tensor pinverse(const Tensor& self, double rcond) {
...
+Tensor& linalg_matrix_rank_out(const Tensor& input, optional<double> tol, bool hermitian, Tensor& result) {
+  double tol_value;
+  Tensor atol, rtol;
   if (tol.has_value()) {
-    double tol_value = tol.value();
-    at::sum_out(result, S > tol_value, /*dim=*/-1);
+    tol_value = tol.value();
+    // For NumPy compatibility tol is not scaled with max(singular_value) if the value for tol is provided
+    // It is assumed that the provided value is the absolute tolerance
+    atol = at::full({}, tol_value, input.options().dtype(ScalarType::Double));
```
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:268]: What about casting to input's dtype?

File: test/test_linalg.py [hunk 3]
```diff
@@ -3385,6 +3385,33 @@ def run_test(shape0, shape1, batch):
         for (shape0, shape1), batch in zip(itertools.product(shapes, reversed(shapes)), batches):
             run_test(shape0, shape1, batch)
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_matrix_rank_tol(self, device, dtype):
+
+        def run_test_tol(shape0, shape1, batch):
+            a = torch.randn(*batch, shape0, shape1, dtype=dtype, device=device)
```
Comment (mruberry) [inline test/test_linalg.py:3394]: prefer make_tensor() to torch.randn() in the test suite

File: test/test_linalg.py [hunk 4]
```diff
@@ -3385,6 +3385,33 @@ def run_test(shape0, shape1, batch):
...
+            # Check against NumPy output
+            # Test float tol, and specific value for each matrix
+            tolerances = [float(torch.rand(1)), ]
+            # Test different types of tol tensor
+            for tol_type in all_types():
+                tolerances.append(torch.rand(a.shape[:-2], dtype=torch.double, device=device).to(tol_type))
```
Comment (mruberry) [inline test/test_linalg.py:3400]: make_tensor() here, too

File: torch/linalg/__init__.py [hunk 5]
```diff
@@ -484,7 +484,9 @@
 Args:
     input (Tensor): the input matrix of size `(m, n)` or the batch of matrices of size `(*, m, n)`
                     where `*` is one or more batch dimensions.
-    tol (float, optional): the tolerance value. Default is ``None``
+    tol (float, Tensor, optional): the tolerance value to determine the cutoff for small singular values.
```
Comment (mruberry) [inline torch/linalg/__init__.py:487]: "the tolerance value or values to determine..."

what happens to singular values below their corresponding tolerance?

**Review by mruberry — COMMENTED** [review]
Hey @IvanYashchuk, just a few comments. Is this the best time to add an OpInfo for torch.linalg.matrix_rank?

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/LinearAlgebra.cpp:213]: We discussed this in `torch.linalg.pinv` PR. The decision there was that we shouldn't restrict the type of tolerances if it works as is.
> If directly atol/rtol tensor is passed, then all types should be valid that allow multiplication with max_singular_value tensor (that can be only of type float or double) and that allow comparison with a tensor of type float or double. So I guess any floating and integer types (so everything that is not complex) should be valid here for any allowed input?

See https://github.com/pytorch/pytorch/pull/48399#discussion_r544480613

`test_matrix_rank_tol` demonstrates that types from `torch.testing.all_types()` work for the tolerances tensor.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 2 above)
Comment (IvanYashchuk) [reply aten/src/ATen/native/LinearAlgebra.cpp:268]: We get a number of type `double` from Python, I don't think we should cast it to input's dtype. There are separate. And in PyTorch it's possible to compare tensors with different types.
Same discussion for `torch.linalg.pinv`: https://github.com/pytorch/pytorch/pull/48399#discussion_r544443532

**Comment by IvanYashchuk** [comment]
Thank you, Mike! Sure, I will add an OpInfo to this PR.

File: torch/linalg/__init__.py (hunk 5 above)
Comment (IvanYashchuk) [reply torch/linalg/__init__.py:487]: They're not considered in the computation of matrix rank then. It's described in the main text:
"The matrix rank is computed as the number of singular values that are greater than the specified tol threshold."
I don't think we need to duplicate this sentence here, right?

Comment (mruberry) [reply torch/linalg/__init__.py:546]: I agree with you.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/native/LinearAlgebra.cpp:218]: Sounds good. It's nice to know we're consistent ;)

**Comment by IvanYashchuk** [comment]
@mruberry I added OpInfo. I think this PR is ready for merge.

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -8855,11 +8855,19 @@
...
   variants: function
 
-- func: linalg_matrix_rank.out(Tensor self, float? tol=None, bool hermitian=False, *, Tensor(a!) out) -> Tensor(a!)
+- func: linalg_matrix_rank.tol_tensor(Tensor input, Tensor tol, bool hermitian=False) -> Tensor
+  python_module: linalg
+  variants: function
+
+- func: linalg_matrix_rank.out(Tensor input, float? tol=None, bool hermitian=False, *, Tensor(a!) out) -> Tensor(a!)
+  python_module: linalg
+  variants: function
```
Comment (ezyang) [inline aten/src/ATen/native/native_functions.yaml:8868]: Doesn't have to be this PR, but consider making this a structured function! https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

Comment (IvanYashchuk) [reply aten/src/ATen/native/native_functions.yaml:8868]: Sure, I think we will port all linear algebra functions to structured kernels after 1.9 release.

**Review by mruberry — APPROVED** [review]
Sorry about the delay; thanks for the ping, @IvanYashchuk

**Pull request closed without merging**